[WIP] web dataset prototype

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48849 webdataset ExpandUrlMacroIterDataset protype
* #48295 [WIP] add RoutedDecoderIterDataset prototype
* **#47831 [WIP] web dataset prototype**

This is a prototype code for new web dataset, we should have the following components:

- [x] ListDirFilesIterDataset  - read file pathname from root and yield file stream
- [x] LoadFilesFromDiskIterDataset - load files' binary and yield steam of (pathname, binary stream), support extraction.
- [x] ExtractFilesIterDataset - unzip tar/zip from a dataset of (pathname, binary). Yield steams of unzipped (pathname, binary)
- [x] RoutedDecoderIterDataset - support txt,json,png, and more.
- [ ] GroupByFilenameIterDataset - TODO: depend on above
- [ ] MapFunctionIterDataSet - TODO: future work maybe
- [ ] Map version of above - TODO: Need to discuss what should be the label.


example:
```
In [1]: from torch.utils.data._dataset import *

In [2]: a = ListDirFilesIterDataset('./torch/utils/data', '*.tar*')

In [3]: for x in a:
   ...:     print(x)
   ...: 
./torch/utils/data/test.tar.gz
./torch/utils/data/outer.test.tar.gz

In [4]: b = LoadFilesFromDiskIterDataset(a, auto_extract=False)

In [5]: len(b)
Out[5]: 2

In [6]: b = LoadFilesFromDiskIterDataset(a, auto_extract=True)

In [7]: len(b)
Out[7]: 9

In [8]: for x in b:
   ...:     print(x[0])
   ...: 
torch/utils/data/test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/dataloader.py
torch/utils/data/outer.test.tar.gz/_dataset.py
torch/utils/data/outer.test.tar.gz/dataset.py
torch/utils/data/outer.test.tar.gz/distributed.py
torch/utils/data/outer.test.tar.gz/__init__.py
torch/utils/data/outer.test.tar.gz/sampler.py
torch/utils/data/outer.test.tar.gz/test.tar.gz/common.py

In [9]: b = LoadFilesFromDiskIterDataset(a, auto_extract=False)

In [10]: c = ExtractFilesIterDataset(b, recursive=False)

In [11]: for x in c:
    ...:     print(x[0])
    ...: 
torch/utils/data/test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/dataloader.py
torch/utils/data/outer.test.tar.gz/_dataset.py
torch/utils/data/outer.test.tar.gz/dataset.py
torch/utils/data/outer.test.tar.gz/distributed.py
torch/utils/data/outer.test.tar.gz/__init__.py
torch/utils/data/outer.test.tar.gz/sampler.py
torch/utils/data/outer.test.tar.gz/test.tar.gz

In [12]: c = ExtractFilesIterDataset(b, recursive=True)

In [13]: for x in c:
    ...:     print(x[0])
    ...: 
torch/utils/data/test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/dataloader.py
torch/utils/data/outer.test.tar.gz/_dataset.py
torch/utils/data/outer.test.tar.gz/dataset.py
torch/utils/data/outer.test.tar.gz/distributed.py
torch/utils/data/outer.test.tar.gz/__init__.py
torch/utils/data/outer.test.tar.gz/sampler.py
torch/utils/data/outer.test.tar.gz/test.tar.gz/common.py

In [14]: 
```

Differential Revision: [D24914126](https://our.internmc.facebook.com/intern/diff/D24914126)

Commit:
web dataset prototype

[ghstack-poisoned]

Commit:
Update on "[WIP] web dataset prototype"


This is a prototype code for new web dataset, we should have the following components:

- [x] ListDirFilesIterDataset  - read file pathname from root
- [x] LoadFilesFromDiskIterDataset - load files' binary and into a list of tuple(pathname, binary), support extraction.
- [x] ExtractFilesIterDataset - extract files from a list of tuple(pathname, binary). Return a list of tuple(pathname, binary)
- [ ] RoutedDecoderIterDataset - TODO: need to discuss the way to support coder, where to get it, etc.
- [ ] GroupByFilenameIterDataset - TODO: depend on above
- [ ] MapFunctionIterDataSet - TODO: future work maybe
- [ ] Map version of above - TODO: Need to discuss what should be the label.

Differential Revision: [D24914126](https://our.internmc.facebook.com/intern/diff/D24914126)

[ghstack-poisoned]

Commit:
Update on "[WIP] web dataset prototype"


This is a prototype code for new web dataset, we should have the following components:

- [x] ListDirFilesIterDataset  - read file pathname from root
- [x] LoadFilesFromDiskIterDataset - load files' binary and into a list of tuple(pathname, binary), support extraction.
- [x] ExtractFilesIterDataset - extract tar/zip from a list of tuple(pathname, binary). Return a list of tuple(pathname, binary)
- [ ] RoutedDecoderIterDataset - TODO: need to discuss the way to support coder, where to get it, etc.
- [ ] GroupByFilenameIterDataset - TODO: depend on above
- [ ] MapFunctionIterDataSet - TODO: future work maybe
- [ ] Map version of above - TODO: Need to discuss what should be the label.

example:
```
In [1]: from torch.utils.data._dataset import *

In [2]: a = ListDirFilesIterDataset('./torch/utils/data', '*.tar*')

In [3]: for x in a:
   ...:     print(x)
   ...: 
./torch/utils/data/test.tar.gz
./torch/utils/data/outer.test.tar.gz

In [4]: b = LoadFilesFromDiskIterDataset(a, auto_extract=False)

In [5]: len(b)
Out[5]: 2

In [6]: b = LoadFilesFromDiskIterDataset(a, auto_extract=True)

In [7]: len(b)
Out[7]: 9

In [8]: for x in b:
   ...:     print(x[0])
   ...: 
torch/utils/data/test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/dataloader.py
torch/utils/data/outer.test.tar.gz/_dataset.py
torch/utils/data/outer.test.tar.gz/dataset.py
torch/utils/data/outer.test.tar.gz/distributed.py
torch/utils/data/outer.test.tar.gz/__init__.py
torch/utils/data/outer.test.tar.gz/sampler.py
torch/utils/data/outer.test.tar.gz/test.tar.gz/common.py

In [9]: b = LoadFilesFromDiskIterDataset(a, auto_extract=False)

In [10]: c = ExtractFilesIterDataset(b, recursive=False)

In [11]: for x in c:
    ...:     print(x[0])
    ...: 
torch/utils/data/test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/dataloader.py
torch/utils/data/outer.test.tar.gz/_dataset.py
torch/utils/data/outer.test.tar.gz/dataset.py
torch/utils/data/outer.test.tar.gz/distributed.py
torch/utils/data/outer.test.tar.gz/__init__.py
torch/utils/data/outer.test.tar.gz/sampler.py
torch/utils/data/outer.test.tar.gz/test.tar.gz

In [12]: c = ExtractFilesIterDataset(b, recursive=True)

In [13]: for x in c:
    ...:     print(x[0])
    ...: 
torch/utils/data/test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/dataloader.py
torch/utils/data/outer.test.tar.gz/_dataset.py
torch/utils/data/outer.test.tar.gz/dataset.py
torch/utils/data/outer.test.tar.gz/distributed.py
torch/utils/data/outer.test.tar.gz/__init__.py
torch/utils/data/outer.test.tar.gz/sampler.py
torch/utils/data/outer.test.tar.gz/test.tar.gz/common.py

In [14]: 
```

Differential Revision: [D24914126](https://our.internmc.facebook.com/intern/diff/D24914126)

[ghstack-poisoned]

Commit:
Update on "[WIP] web dataset prototype"


This is a prototype code for new web dataset, we should have the following components:

- [x] ListDirFilesIterDataset  - read file pathname from root
- [x] LoadFilesFromDiskIterDataset - load files' binary and into a list of tuple(pathname, binary), support extraction.
- [x] ExtractFilesIterDataset - extract tar/zip from a list of tuple(pathname, binary). Return a list of tuple(pathname, binary)
- [ ] RoutedDecoderIterDataset - TODO: need to discuss the way to support coder, where to get it, etc.
- [ ] GroupByFilenameIterDataset - TODO: depend on above
- [ ] MapFunctionIterDataSet - TODO: future work maybe
- [ ] Map version of above - TODO: Need to discuss what should be the label.

example:
```
In [1]: from torch.utils.data._dataset import *

In [2]: a = ListDirFilesIterDataset('./torch/utils/data', '*.tar*')

In [3]: for x in a:
   ...:     print(x)
   ...: 
./torch/utils/data/test.tar.gz
./torch/utils/data/outer.test.tar.gz

In [4]: b = LoadFilesFromDiskIterDataset(a, auto_extract=False)

In [5]: len(b)
Out[5]: 2

In [6]: b = LoadFilesFromDiskIterDataset(a, auto_extract=True)

In [7]: len(b)
Out[7]: 9

In [8]: for x in b:
   ...:     print(x[0])
   ...: 
torch/utils/data/test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/dataloader.py
torch/utils/data/outer.test.tar.gz/_dataset.py
torch/utils/data/outer.test.tar.gz/dataset.py
torch/utils/data/outer.test.tar.gz/distributed.py
torch/utils/data/outer.test.tar.gz/__init__.py
torch/utils/data/outer.test.tar.gz/sampler.py
torch/utils/data/outer.test.tar.gz/test.tar.gz/common.py

In [9]: b = LoadFilesFromDiskIterDataset(a, auto_extract=False)

In [10]: c = ExtractFilesIterDataset(b, recursive=False)

In [11]: for x in c:
    ...:     print(x[0])
    ...: 
torch/utils/data/test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/dataloader.py
torch/utils/data/outer.test.tar.gz/_dataset.py
torch/utils/data/outer.test.tar.gz/dataset.py
torch/utils/data/outer.test.tar.gz/distributed.py
torch/utils/data/outer.test.tar.gz/__init__.py
torch/utils/data/outer.test.tar.gz/sampler.py
torch/utils/data/outer.test.tar.gz/test.tar.gz

In [12]: c = ExtractFilesIterDataset(b, recursive=True)

In [13]: for x in c:
    ...:     print(x[0])
    ...: 
torch/utils/data/test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/dataloader.py
torch/utils/data/outer.test.tar.gz/_dataset.py
torch/utils/data/outer.test.tar.gz/dataset.py
torch/utils/data/outer.test.tar.gz/distributed.py
torch/utils/data/outer.test.tar.gz/__init__.py
torch/utils/data/outer.test.tar.gz/sampler.py
torch/utils/data/outer.test.tar.gz/test.tar.gz/common.py

In [14]: 
```

Differential Revision: [D24914126](https://our.internmc.facebook.com/intern/diff/D24914126)

[ghstack-poisoned]

Commit:
Update on "[WIP] web dataset prototype"


This is a prototype code for new web dataset, we should have the following components:

- [x] ListDirFilesIterDataset  - read file pathname from root
- [x] LoadFilesFromDiskIterDataset - load files' binary and into a list of tuple(pathname, binary), support extraction.
- [x] ExtractFilesIterDataset - extract tar/zip from a list of tuple(pathname, binary). Return a list of tuple(pathname, binary)
- [ ] RoutedDecoderIterDataset - TODO: need to discuss the way to support coder, where to get it, etc.
- [ ] GroupByFilenameIterDataset - TODO: depend on above
- [ ] MapFunctionIterDataSet - TODO: future work maybe
- [ ] Map version of above - TODO: Need to discuss what should be the label.

example:
```
In [1]: from torch.utils.data._dataset import *

In [2]: a = ListDirFilesIterDataset('./torch/utils/data', '*.tar*')

In [3]: for x in a:
   ...:     print(x)
   ...: 
./torch/utils/data/test.tar.gz
./torch/utils/data/outer.test.tar.gz

In [4]: b = LoadFilesFromDiskIterDataset(a, auto_extract=False)

In [5]: len(b)
Out[5]: 2

In [6]: b = LoadFilesFromDiskIterDataset(a, auto_extract=True)

In [7]: len(b)
Out[7]: 9

In [8]: for x in b:
   ...:     print(x[0])
   ...: 
torch/utils/data/test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/dataloader.py
torch/utils/data/outer.test.tar.gz/_dataset.py
torch/utils/data/outer.test.tar.gz/dataset.py
torch/utils/data/outer.test.tar.gz/distributed.py
torch/utils/data/outer.test.tar.gz/__init__.py
torch/utils/data/outer.test.tar.gz/sampler.py
torch/utils/data/outer.test.tar.gz/test.tar.gz/common.py

In [9]: b = LoadFilesFromDiskIterDataset(a, auto_extract=False)

In [10]: c = ExtractFilesIterDataset(b, recursive=False)

In [11]: for x in c:
    ...:     print(x[0])
    ...: 
torch/utils/data/test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/dataloader.py
torch/utils/data/outer.test.tar.gz/_dataset.py
torch/utils/data/outer.test.tar.gz/dataset.py
torch/utils/data/outer.test.tar.gz/distributed.py
torch/utils/data/outer.test.tar.gz/__init__.py
torch/utils/data/outer.test.tar.gz/sampler.py
torch/utils/data/outer.test.tar.gz/test.tar.gz

In [12]: c = ExtractFilesIterDataset(b, recursive=True)

In [13]: for x in c:
    ...:     print(x[0])
    ...: 
torch/utils/data/test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/dataloader.py
torch/utils/data/outer.test.tar.gz/_dataset.py
torch/utils/data/outer.test.tar.gz/dataset.py
torch/utils/data/outer.test.tar.gz/distributed.py
torch/utils/data/outer.test.tar.gz/__init__.py
torch/utils/data/outer.test.tar.gz/sampler.py
torch/utils/data/outer.test.tar.gz/test.tar.gz/common.py

In [14]: 
```

Differential Revision: [D24914126](https://our.internmc.facebook.com/intern/diff/D24914126)

[ghstack-poisoned]

glaringlee commented:
This is now totally streaming based datasets.

ejguan commented:
I thought we should support shardable Dataset at most of levels and hide within the pipeline.
For the developer who will be working on the Dataset, we need to expose a sharding_fn to let them shard. And, we can abstract a higher level Dataset like `Webdataset` or whatever, and use our own sharding strategy then provide the abstract Dataset for user. 
So, we may also add an argument for sharding (as well as seeding...).

glaringlee commented:
> I thought we should support shardable Dataset at most of levels and hide within the pipeline.
> For the developer who will be working on the Dataset, we need to expose a sharding_fn to let them shard. And, we can abstract a higher level Dataset like `Webdataset` or whatever, and use our own sharding strategy then provide the abstract Dataset for user.
> So, we may also add an argument for sharding (as well as seeding...).

Yep, agree. sharding_fn will be provided later, was making the layered streaming works first.  This is not a 'web' dataset at the moment, it can only handle local files, the functions in common.py already took care of PIPE'ed stream which can be used in webdataset with no change. I am going to add route dataset for decoding, and will then go back on sharding. We can discuss further.

Commit:
Update on "[WIP] web dataset prototype"


This is a prototype code for new web dataset, we should have the following components:

- [x] ListDirFilesIterDataset  - read file pathname from root and yield file stream
- [x] LoadFilesFromDiskIterDataset - load files' binary and yield steam of (pathname, binary stream), support extraction.
- [x] ExtractFilesIterDataset - unzip tar/zip from a dataset of (pathname, binary). Yield steams of unzipped (pathname, binary)
- [ ] RoutedDecoderIterDataset - TODO: need to discuss the way to support coder, where to get it, etc.
- [ ] GroupByFilenameIterDataset - TODO: depend on above
- [ ] MapFunctionIterDataSet - TODO: future work maybe
- [ ] Map version of above - TODO: Need to discuss what should be the label.

example:
```
In [1]: from torch.utils.data._dataset import *

In [2]: a = ListDirFilesIterDataset('./torch/utils/data', '*.tar*')

In [3]: for x in a:
   ...:     print(x)
   ...: 
./torch/utils/data/test.tar.gz
./torch/utils/data/outer.test.tar.gz

In [4]: b = LoadFilesFromDiskIterDataset(a, auto_extract=False)

In [5]: len(b)
Out[5]: 2

In [6]: b = LoadFilesFromDiskIterDataset(a, auto_extract=True)

In [7]: len(b)
Out[7]: 9

In [8]: for x in b:
   ...:     print(x[0])
   ...: 
torch/utils/data/test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/dataloader.py
torch/utils/data/outer.test.tar.gz/_dataset.py
torch/utils/data/outer.test.tar.gz/dataset.py
torch/utils/data/outer.test.tar.gz/distributed.py
torch/utils/data/outer.test.tar.gz/__init__.py
torch/utils/data/outer.test.tar.gz/sampler.py
torch/utils/data/outer.test.tar.gz/test.tar.gz/common.py

In [9]: b = LoadFilesFromDiskIterDataset(a, auto_extract=False)

In [10]: c = ExtractFilesIterDataset(b, recursive=False)

In [11]: for x in c:
    ...:     print(x[0])
    ...: 
torch/utils/data/test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/dataloader.py
torch/utils/data/outer.test.tar.gz/_dataset.py
torch/utils/data/outer.test.tar.gz/dataset.py
torch/utils/data/outer.test.tar.gz/distributed.py
torch/utils/data/outer.test.tar.gz/__init__.py
torch/utils/data/outer.test.tar.gz/sampler.py
torch/utils/data/outer.test.tar.gz/test.tar.gz

In [12]: c = ExtractFilesIterDataset(b, recursive=True)

In [13]: for x in c:
    ...:     print(x[0])
    ...: 
torch/utils/data/test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/dataloader.py
torch/utils/data/outer.test.tar.gz/_dataset.py
torch/utils/data/outer.test.tar.gz/dataset.py
torch/utils/data/outer.test.tar.gz/distributed.py
torch/utils/data/outer.test.tar.gz/__init__.py
torch/utils/data/outer.test.tar.gz/sampler.py
torch/utils/data/outer.test.tar.gz/test.tar.gz/common.py

In [14]: 
```

Differential Revision: [D24914126](https://our.internmc.facebook.com/intern/diff/D24914126)

[ghstack-poisoned]

Commit:
Update on "[WIP] web dataset prototype"


This is a prototype code for new web dataset, we should have the following components:

- [x] ListDirFilesIterDataset  - read file pathname from root and yield file stream
- [x] LoadFilesFromDiskIterDataset - load files' binary and yield steam of (pathname, binary stream), support extraction.
- [x] ExtractFilesIterDataset - unzip tar/zip from a dataset of (pathname, binary). Yield steams of unzipped (pathname, binary)
- [x] RoutedDecoderIterDataset - support txt,json,png, and more.
- [ ] GroupByFilenameIterDataset - TODO: depend on above
- [ ] MapFunctionIterDataSet - TODO: future work maybe
- [ ] Map version of above - TODO: Need to discuss what should be the label.


example:
```
In [1]: from torch.utils.data._dataset import *

In [2]: a = ListDirFilesIterDataset('./torch/utils/data', '*.tar*')

In [3]: for x in a:
   ...:     print(x)
   ...: 
./torch/utils/data/test.tar.gz
./torch/utils/data/outer.test.tar.gz

In [4]: b = LoadFilesFromDiskIterDataset(a, auto_extract=False)

In [5]: len(b)
Out[5]: 2

In [6]: b = LoadFilesFromDiskIterDataset(a, auto_extract=True)

In [7]: len(b)
Out[7]: 9

In [8]: for x in b:
   ...:     print(x[0])
   ...: 
torch/utils/data/test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/dataloader.py
torch/utils/data/outer.test.tar.gz/_dataset.py
torch/utils/data/outer.test.tar.gz/dataset.py
torch/utils/data/outer.test.tar.gz/distributed.py
torch/utils/data/outer.test.tar.gz/__init__.py
torch/utils/data/outer.test.tar.gz/sampler.py
torch/utils/data/outer.test.tar.gz/test.tar.gz/common.py

In [9]: b = LoadFilesFromDiskIterDataset(a, auto_extract=False)

In [10]: c = ExtractFilesIterDataset(b, recursive=False)

In [11]: for x in c:
    ...:     print(x[0])
    ...: 
torch/utils/data/test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/dataloader.py
torch/utils/data/outer.test.tar.gz/_dataset.py
torch/utils/data/outer.test.tar.gz/dataset.py
torch/utils/data/outer.test.tar.gz/distributed.py
torch/utils/data/outer.test.tar.gz/__init__.py
torch/utils/data/outer.test.tar.gz/sampler.py
torch/utils/data/outer.test.tar.gz/test.tar.gz

In [12]: c = ExtractFilesIterDataset(b, recursive=True)

In [13]: for x in c:
    ...:     print(x[0])
    ...: 
torch/utils/data/test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/common.py
torch/utils/data/outer.test.tar.gz/dataloader.py
torch/utils/data/outer.test.tar.gz/_dataset.py
torch/utils/data/outer.test.tar.gz/dataset.py
torch/utils/data/outer.test.tar.gz/distributed.py
torch/utils/data/outer.test.tar.gz/__init__.py
torch/utils/data/outer.test.tar.gz/sampler.py
torch/utils/data/outer.test.tar.gz/test.tar.gz/common.py

In [14]: 
```

Differential Revision: [D24914126](https://our.internmc.facebook.com/intern/diff/D24914126)

[ghstack-poisoned]

VitalyFedyunin commented on torch/utils/data/common.py:
```diff
@@ -0,0 +1,162 @@
+        binary_stream : Any,
+        extract_tars : bool = True,
+        extract_zips : bool = True,
+        recursive : bool = True,
```
Please no recursive tar/zip files =) 🤯

VitalyFedyunin commented on torch/utils/data/common.py:
```diff
@@ -0,0 +1,162 @@
+                        yield (inner_pathname, extract_fobj)
+            return
+        except tarfile.TarError:
+            if not seekable:
```
Please raise Exception in this case, we don't want to silently swallow errors in care of bad files

VitalyFedyunin commented on torch/utils/data/common.py:
```diff
@@ -0,0 +1,162 @@
+                return
+            binary_stream.seek(0)
+
+    yield (pathname, binary_stream)
```
Wait, here you will yield input stream AFTER yielding all unpacked files, please remove this part.

glaringlee replied:
I think this is as designed. We will only touch this piece when input steam is not a supported zip format. (not tar/zip), and we return the original binary stream in this case.

VitalyFedyunin reviewed (commented):
More comments to go...

glaringlee commented:
moved to #48850 due to stack messed up, I'll keep this open for a while to resolve all comments.

glaringlee commented:
moved to #48944

harpone commented:
How does all this relate to @tmbdev 's elegant webdataset solution? https://github.com/tmbdev/webdataset

What are the key differences/similarities? I'm using @tmbdev webdataset now and it would be nice to know something of the future direction of both versions...

glaringlee commented:
All of these is to support the new data loader in pytorch. Please see this doc.
https://github.com/pytorch/pytorch/issues/49440

We are still iterating the design, but basically the difference between tmbdev's webdataset and this one is that we will decouple all the components within dataset. (sampler, shuffler, etc.) Each component will be treated as a separate layer named data pipe. In the future, a webdataset will be a factory of data pipe. For eg.
ParseURLDatapipe->DownloadWebDatapipe->TarDatapipe->SamplerDatapipe->ShufflerDatapipe->......

harpone commented:
@glaringlee ah OK! Ran a search "webdataset" in the issues and missed the docs because there's no mention of webdataset... I'll take a look. Thanks for the info!

tmbdev commented:
> but basically the difference between tmbdev's webdataset and this one is that we will decouple all the components within dataset.

@glaringlee That's really not a difference. The components in WebDataset are "decoupled" already.

> ParseURLDatapipe->DownloadWebDatapipe->TarDatapipe->SamplerDatapipe->ShufflerDatapipe->......

The base pipeline in WebDataset is ShardList->url_opener->cache_shards->tar_file_expander->group_by_keys; there is a convenience function that creates this for you in one step, but if you like, you can also just create the components separately.

You can add shuffling, batching, decoding, etc. onto that as needed.

glaringlee commented:
@tmbdev 
Ah, thanks for updating. I did not check the webdataset repo since last Dec. The webdataset was not decoupled when I wrote this PR on last Nov. And yes, if webdataset is already decoupled, then there should be no much difference on high level design. 

This is a deprecated PR by the way, our latest implementation is in torch.data.utils.datapipes at the moment, more will be landed soon.  I am currently using the decoder that created from webdataset :)
